### PR TITLE
Enhance Lcov format parse functionaility

### PR DESCRIPTION
**Pull request by 129tyc**
This pull request introduces support for parsing branch and function coverage data in LCOV reports, enhances the filtering logic for coverage analysis, and updates the test suite to validate the new functionality. The most important changes include modifications to the `parse` function to handle branch and function data, updates to test fixtures, and the addition of new test cases to ensure correctness.

### Enhancements to LCOV Parsing:
* Added support for parsing branch coverage (`BRDA`) and function coverage (`FN`, `FNDA`) directives in the `parse` function. This includes tracking branch hit/miss data and associating hit counts with functions. [[1]](diffhunk://#diff-08d5c38222d19faa735e78af8c8d0148c451ce7a7378240cdc7c999faef486acR313-R314) [[2]](diffhunk://#diff-08d5c38222d19faa735e78af8c8d0148c451ce7a7378240cdc7c999faef486acR341-R396)
* Introduced logic to filter out lines with partial branch coverage and to apply function hit filtering, ensuring only fully covered branches and hit functions are considered.

### Updates to Test Fixtures:
* Updated the `add_console_report_lcov.txt` test fixture to reflect the new branch and function coverage reporting.
* Adjusted the `lcov.info` test fixture to align with the new parsing logic for function coverage.

### Test Suite Enhancements:
* Replaced an existing test reference to use the updated `add_console_report_lcov.txt` fixture.
* Added a new test case, `test_parse_branch_and_function_coverage`, to validate the parsing and filtering of branch and function coverage data. This ensures the correct behavior of the updated `parse` function.

These changes collectively improve the accuracy and granularity of coverage analysis while ensuring robust validation through comprehensive test coverage.

### CODE REVIEW

File: diff_cover/violationsreporters/violations_reporter.py [hunk 1]
```diff
@@ -336,22 +338,62 @@ def parse(lcov_file):
                 if line_no not in lcov_report[source_file]:
                     lcov_report[source_file][line_no] = 0
                 lcov_report[source_file][line_no] += num_executions
+            elif directive == "BRDA":
+                args = content.split(",")
+                if len(args) != 4:
+                    continue
+                line_no = int(args[0])
```
Comment (kingbuzzman) [inline diff_cover/violationsreporters/violations_reporter.py:345]: Again cue from above
```suggestion
                if source_file is None:
                    raise ValueError(
                        f"No source file specified for line coverage: {line}"
                    )
                line_no = int(args[0])
```

File: diff_cover/violationsreporters/violations_reporter.py [hunk 2]
```diff
@@ -336,22 +338,62 @@ def parse(lcov_file):
...
+                    branch_coverage[source_file][line_no]["total"] += 1
+                    if int(taken) > 0:
+                        branch_coverage[source_file][line_no]["hit"] += 1
+                        if line_no not in lcov_report[source_file]:
+                            lcov_report[source_file][line_no] = 0
+                        lcov_report[source_file][line_no] += int(taken)
+            elif directive == "FN":
+                args = content.split(",")
+                if len(args) != 2:
+                    continue
```
Comment (kingbuzzman) [inline diff_cover/violationsreporters/violations_reporter.py:357]: Ditto.

File: diff_cover/violationsreporters/violations_reporter.py [hunk 3]
```diff
@@ -336,22 +338,62 @@ def parse(lcov_file):
...
+                line_no = int(args[0])
+                func_name = args[1]
+                function_lines[source_file][func_name] = (line_no, 0)
+            elif directive == "FNDA":
+                args = content.split(",")
+                if len(args) != 2:
+                    continue
```
Comment (kingbuzzman) [inline diff_cover/violationsreporters/violations_reporter.py:364]: Ditto

File: diff_cover/violationsreporters/violations_reporter.py [hunk 4]
```diff
@@ -336,22 +338,62 @@ def parse(lcov_file):
...
+            elif directive == "BRDA":
+                args = content.split(",")
+                if len(args) != 4:
+                    continue
+                line_no = int(args[0])
+                taken = args[3]
+                if taken != "-" and source_file:
+                    branch_coverage[source_file][line_no]["total"] += 1
+                    if int(taken) > 0:
+                        branch_coverage[source_file][line_no]["hit"] += 1
```
Comment (kingbuzzman) [inline diff_cover/violationsreporters/violations_reporter.py:350]: Theres a lot of nesting, can we flatten it out, and throw errors when possible?

**Review by kingbuzzman — CHANGES_REQUESTED** [review]

**Comment by kingbuzzman** [comment]
On the surface, I say i'm good with it...
* Im not a huge fan of the modification of `tests/fixtures/lcov.info` as this is used 3 existing tests -- and frankly it give me pause you didn't have to modify them... which... scares me...
* Im by no means an expert in `lcov` -- as a matter of fact learned about its existence about 2 weeks ago 😇

@Bachmann1234 can you authorize @129tyc to run CI, I'd like to see the test coverage and I would give you a better assessment.

**Comment by 129tyc** [comment]
> On the surface, I say i'm good with it...
> 
> * Im not a huge fan of the modification of `tests/fixtures/lcov.info` as this is used 3 existing tests -- and frankly it give me pause you didn't have to modify them... which... scares me...
> * Im by no means an expert in `lcov` -- as a matter of fact learned about its existence about 2 weeks ago 😇
> 
> @Bachmann1234 can you authorize @129tyc to run CI, I'd like to see the test coverage and I would give you a better assessment.

Thanks for your review!

- I understand your concern about the test file. The reason I modified it is because I personally believe branch lines shouldn’t count as line hits. But I realize that may be valid behavior in some languages, so I’m happy to revert that part and update the assertions in the test.
- Honestly, I felt the same. My original plan was to fix the coverage issues using the XML report format, but after some research, I found that XML reports don’t contain enough detail about branch and function call coverage. LCOV format does. I got help from AI to complete the implementation, so all I could do was trust it. If you think this PR is too risky to merge, I can keep it in my fork and close this one. I just wanted to contribute to the open source community.

**Comment by kingbuzzman** [comment]
I think there's a couple of good things here, i wouldn't throw it out. I do suggest you brush up on lcov some more so you know and can recognize when the AI is hallucinating and giving you garbage. Like i mentioned before, i'm in no way an expert in the format, i'd like to get an existing "real" example and using that as a fixture to ensure it does work.

**Comment by Bachmann1234** [comment]
I cloned the fork and ran the tests since the formatter failing is not terribly interesting. 

They are passing. I am a bit surprised that modifying this file did not break anything. However, I am also not an lcov expert. 

Are you opening this to solve a problem you are having? Or is this an attempt to improve this for fun? I ask because im more comfortable releasing this if I know you are going to use it and be able to report back any issues

**Comment by 129tyc** [comment]
> I cloned the fork and ran the tests since the formatter failing is not terribly interesting.
> 
> They are passing. I am a bit surprised that modifying this file did not break anything. However, I am also not an lcov expert.
> 
> Are you opening this to solve a problem you are having? Or is this an attempt to improve this for fun? I ask because im more comfortable releasing this if I know you are going to use it and be able to report back any issues

Thanks for testing! I opened this PR because I use this tool to calculate diff coverage data for TypeScript at my current company, and I noticed that some coverage records were missing for anonymous functions and branches. I applied these fixes in my forked repo, replaced the original one, and everything is working fine. I like this tool because it’s open source and free, which makes it a great alternative to other third-party services. So yes, I’ll keep working on this feature.

**Comment by 129tyc** [comment]
> @129tyc can you run `black` and commit your code again, I'd like to see some stats on this PR. Thanks!

No problem. Done!

**Comment by kingbuzzman** [comment]
LOL. Well in terms of coverage, you're only missing all the edge cases where you're raising an exception because the file is malformed. If @Bachmann1234 is ok merging it and you keep using/testing it, and identifying any issues, im fine with it 👍

File: tests/fixtures/lcov.info [hunk 5]
```diff
@@ -21,4 +20,4 @@ LH:6
 BRF:0
 BRH:0
 BRDA:3,0,0,0
-end_of_record
\ No newline at end of file
+end_of_record
```
Comment (kingbuzzman) [inline tests/fixtures/lcov.info:23]: Have you had a chance to look at the .lcov life and see if there is a trailing space? I think i originally asked for you to add a trailing space because I thought it was a .py file.. now im not so sure..

File: tests/test_violations_reporter.py [hunk 6]
```diff
@@ -972,25 +972,94 @@ def test_no_such_file(self):
...
+        }
+
+    def _coverage_lcov(
+        self,
+        file_paths,
+        violations,
+        measured,
+        branch_data=None,
+        function_data=None,
+    ):
```
Comment (kingbuzzman) [inline tests/test_violations_reporter.py:1028]: I can make this more pytest-y later.

File: diff_cover/violationsreporters/violations_reporter.py [hunk 7]
```diff
@@ -384,33 +384,48 @@ def parse(lcov_file):
...
+                    has_da_directive = line_no in lcov_report[source_file]
+
+                    if not has_da_directive:
+                        # No line execution data, use branch coverage
+                        if info["total"] > 0 and info["hit"] < info["total"]:
+                            lcov_report[source_file][
+                                line_no
+                            ] = 0  # Partial branch coverage
+                        else:
+                            lcov_report[source_file][line_no] = info["executions"]
```
Comment (kingbuzzman) [inline diff_cover/violationsreporters/violations_reporter.py:414]:
```suggestion
                            lcov_report[source_file][line_no] = info["executions"]
                        continue
```

File: diff_cover/violationsreporters/violations_reporter.py [hunk 8]
```diff
@@ -384,33 +384,48 @@ def parse(lcov_file):
...
+                    elif not lcov_report[source_file][line_no]:
+                        # Line shows 0 executions, but check if branches were hit
+                        if info["executions"] > 0:
+                            lcov_report[source_file][line_no] = info["executions"]
+                    # Note: Don't override existing positive execution counts
```
Comment (kingbuzzman) [inline diff_cover/violationsreporters/violations_reporter.py:419]: I dont see how these blocks are related to each other
```suggestion
                    if not lcov_report[source_file][line_no]:
                        # Line shows 0 executions, but check if branches were hit
                        if info["executions"] > 0:
                            lcov_report[source_file][line_no] = info["executions"]
                    # Note: Don't override existing positive execution counts
```

File: tests/test_integration.py [hunk 9]
```diff
@@ -397,6 +397,79 @@ def test_expand_coverage_report_uncomplete_report(
...
+        assert runbin(["cpp_functions_coverage.lcov"]) == 0
+
+        # Capture and validate output
+        output = capsys.readouterr().out
+
+        # Should handle FNL/FNA directives without errors
+        assert "calculator.cpp" in output
+        assert "Coverage: " in output
+        # Should not crash on FNL/FNA directives
+        assert "Error" not in output and "Exception" not in output
```
Comment (kingbuzzman) [inline tests/test_integration.py:452]: Im not convinced this is the best way to test this

File: tests/test_integration.py [hunk 10]
```diff
@@ -397,6 +397,79 @@ def test_expand_coverage_report_uncomplete_report(
...
+        assert runbin(["typescript_branches_coverage.lcov"]) == 0
+
+        # Capture and validate output
+        output = capsys.readouterr().out
+
+        # Should handle BRDA directives properly
+        assert "calculator.ts" in output
+        assert "Coverage: " in output
+        # Should apply branch coverage logic correctly
+        assert "Error" not in output and "Exception" not in output
```
Comment (kingbuzzman) [inline tests/test_integration.py:471]: Ditto.

**Review by kingbuzzman — APPROVED** [review]
I'm good with this @Bachmann1234 , thank you @129tyc !

**Comment by Bachmann1234** [comment]
@129tyc this is released https://pypi.org/project/diff-cover/9.4.0/
